master/pkg/searcher/asha.go

```diff
@@ -0,0 +1,244 @@
+package searcher
+
+import (
+	"math"
+	"sort"
+
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+// AsyncHalvingSearch implements a search using the asynchronous successive halving algorithm
+// (ASHA). The experiment will run until the target number of trials have been completed
+// in the bottom rung and no further promotions can be made to higher rungs.
+type asyncHalvingSearch struct {
+	defaultSearchMethod
+	model.AsyncHalvingConfig
+
+	rungs      []*rung
+	trialRungs map[RequestID]int
+	// earlyExitTrials contains trials that exited early that are still considered in the search.
+	earlyExitTrials     map[RequestID]bool
+	totalStepsCompleted int
+	maxTrials           int
+	trialsCompleted     int
+}
+
+const ashaExitedMetricValue = math.MaxFloat64
+
+func newAsyncHalvingSearch(config model.AsyncHalvingConfig) SearchMethod {
+	rungs := make([]*rung, 0, config.NumRungs)
+	for id := 0; id < config.NumRungs; id++ {
+		compound := math.Pow(config.Divisor, float64(config.NumRungs-id-1))
+		stepsNeeded := max(int(float64(config.TargetTrialSteps)/compound), 1)
+		rungs = append(rungs,
+			&rung{
+				stepsNeeded:       stepsNeeded,
+				outstandingTrials: 0})
+	}
+
+	return &asyncHalvingSearch{
+		AsyncHalvingConfig: config,
+		rungs:              rungs,
+		trialRungs:         make(map[RequestID]int),
+		earlyExitTrials:    make(map[RequestID]bool),
+		maxTrials:          config.MaxTrials,
+	}
+}
+
+// promotions handles bookkeeping of validation metrics and returns a RequestID to promote if
+// appropriate.
+func (r *rung) promotionsAsync(requestID RequestID, metric float64, divisor float64) []RequestID {
+	// See if there is a trial to promote. We are increasing the total number of trials seen by 1; the
+	// number of best trials that definitely should have been promoted so far (numPromote) can only
+	// stay the same or increase by 1.
+	oldNumPromote := int(float64(len(r.metrics)) / divisor)
+	numPromote := int(float64(len(r.metrics)+1) / divisor)
+
+	// Insert the new trial result in the appropriate place in the sorted list.
+	insertIndex := sort.Search(
+		len(r.metrics),
+		func(i int) bool { return r.metrics[i].metric > metric },
+	)
+	promoteNow := insertIndex < numPromote
+
+	r.metrics = append(r.metrics, trialMetric{})
+	copy(r.metrics[insertIndex+1:], r.metrics[insertIndex:])
+	r.metrics[insertIndex] = trialMetric{
+		requestID: requestID,
+		metric:    metric,
+		promoted:  promoteNow,
+		closed:    false,
+	}
+
+	// If the new trial is good enough, it should be promoted immediately (whether or not numPromote
+	// changes). Otherwise, if numPromote changes, there is some other trial that should be promoted,
+	// unless it has been promoted already.
+	if promoteNow {
+		return []RequestID{requestID}
+	}
+	if numPromote != oldNumPromote {
+		t := &r.metrics[oldNumPromote]
+		if !t.promoted {
+			t.promoted = true
+			return []RequestID{t.requestID}
+		}
```

Comment:
question: what is the situation where this trial is already promoted and we didn't have a `promoteNow` I'm having trouble thinking through why this would be the case.

Reply:
trial 1: 0.2
trial 2: 0.3
trial 3: 0.4
trial 4: 0.5
-> promote 1
trial 5: 0.1
-> promote 5
trial 6: 0.6
trial 7: 0.7
trial 8: 0.8
-> nothing to promote even though oldNumPromote < numPromote because trials 1 and 5 already promoted

Reply:
the idea here is that we promoted `trial 1` early. So `oldNumPromote` was `1` when we promoted trial 5?

Reply:
yes, so when trial 5 finishes, even though oldNumPromote remains 1, the insert index of trial 5 is 0 so we will promote it now even though we've promoted trial 1 already.

Reply:
cool -- and in an ideal world would we preempt trial 1 or do you let them both train and see what happens?

Reply:
we let them both train.  in the limit the incorrect promotions go to zero and also the incorrect promotions will only get promoted further if they continue to outperform, meaning we should have promoted it earlier even if relative performance at a given rung was bad.